File: branches/scala_implementation/taboosearch/src/taboosearch/tsp/TspSolution.scala
package taboosearch.tsp

import scala.collection.immutable
import core._
// http://www.scala-lang.org/api/current/scala/collection/immutable/IndexedSeq.html


/* "Ленивое" решение задачи коммивояжёра. Представляет собой пару (route, move),
 * где route -- маршрут коммивояжера, move -- индексы вершин, которые необходимо
 * помять местами в route.
 */
class TspSolution(val route : List[Int], val move : Pair[Int, Int])
							extends Solution with immutable.IndexedSeq[Int] {

	// Конструктор вида TspSolution(1, 2, 3, 4, 5)
	def this(route : Int*) = this(route.toList, null)
	
	// Конструктор вида TspSolution(List(1, 2, 3, 4, 5))
	def this(route : List[Int]) = this(route, null)
	
	override def apply(idx : Int) = route(idx % route.length) // абстрактный метод в IndexedSeq
	
	override def length() = route.length // абстрактный метод в IndexedSeq
	
	override def copy() : TspSolution = new TspSolution(route, move);
	
	override def toString() : String = "[" + unlazify().mkString(", ") + " ]"
	
	def unlazify() : List[Int] = {
		if (move != null) {
			val routeCopy = route.toArray
			val (i, j) = move;		
			val tmp = routeCopy(i);
			routeCopy(i) = routeCopy(j);
			routeCopy(j) = tmp;
			return routeCopy.toList;
		} else {
			return route;
		}
	}
	
}

File: branches/new_interfaces/src/taboosearch/tsp/TspSolution.java
package taboosearch.tsp;

public class TspSolution extends taboosearch.Solution {
	// immutable
	final private TspSalesmanRoute route;
	final private double routeCost;
	final private TspSwapMove move;

	public TspSolution(TspSalesmanRoute route, double routeCost, TspSwapMove move) {
		this.route = route;
		this.routeCost = routeCost;
		this.move = move;
	}
	
	public TspSolution(TspSalesmanRoute route, double routeCost) {
		this(route, routeCost, null);
	}
	
	public TspSalesmanRoute getRoute() {
		return this.route;
	}
	
	public double getRouteCost() {
		return this.routeCost;
	}
	
	public TspSwapMove getMove() {
		return move;
	}
	
	public boolean isLazy() {
		return this.move != null;
	}

	public TspSalesmanRoute unlazify() {
		if (this.move != null) {
			int[] routeArray = this.route.toArray();
			int i = this.move.getI(),
				j = this.move.getJ();
			
			int tmp = routeArray[i];
			routeArray[i] = routeArray[j];
			routeArray[j] = tmp;
			
			return new TspSalesmanRoute(routeArray);
		} else {
			return this.route;
		}
	}

	@Override
	public String getStringRepresentation() {
		StringBuilder result = new StringBuilder();
		result.append("[ ");
		
		TspSalesmanRoute route = this.unlazify();
		result.append(route.get(0) + 1);
		for (int i = 1; i < route.length(); ++i) {
			result.append(", ");
			result.append(route.get(i) + 1);
		}

		result.append(" ] ");
		return result.toString();
	}

	@Override
	public TspSolution copy() {
		return new TspSolution(route, routeCost, move);
	}

}
